### PR TITLE
Add support for in_axes=None (but not out_axes, or in_axes>0) to pmap

### PR DESCRIPTION
**Pull request by jekbradbury**
After trying a few different ways to give XLA information about when arguments to a pmap are the same across replicas (needed to trigger a strong-scaling optimization), this seems like a sweet spot:
- it can use the existing `mapped_invars` param, so there are no changes needed (yet) on the `xla_pmap` primitive or transformations
- it's (half of) the API I think we want in the long run, rather than an experimental keyword argument we'll delete later
- it's already useful for "normal" users (using `in_axes=None` for a replicated `DeviceArray` pmap input rather than manually broadcasting will avoid the relatively slow `DeviceArray` indexing path #2891 for that argument; this would help with #2871)

After this change, there are two ways to use pmap with replicated arguments that are also return values (e.g. neural network parameters in data-parallel training):
- most users should keep using `in_axes=0` (the default), leading to device persistent `ShardedDeviceArray`s that are sharded along their leading axis
- those of us who need the relevant XLA:TPU optimization can use `in_axes=None`, and manually adjust the `sharding_spec` of the returned `ShardedDeviceArray` to use `replication_factor` instead of leading-axis sharding (keeping this temporary/experimental/ugly code out of the core).

Later on, I also want to support `in_axes`>0 as well as `out_axes`. But there's a lot of implementation subtlety and choices related to `out_axes`, and I don't want to continue blocking things on that.

Once `out_axes=None` works, pmaps that use that will return replicated `ShardedDeviceArrays` with no leading replica axis, enabling the same kind of automatic device persistence behavior for `in_`/`out_axes=None` values that `in_`/`out_axes=0` values already have.

### CODE REVIEW

**Comment by jekbradbury** [comment]
@skye I think this is ready for review 🙂

**Review by mattjj — APPROVED** [review]
LGTM at the API level! This is fantastic.

File: jax/interpreters/pxla.py [hunk 1]
```diff
@@ -196,6 +196,7 @@ def shard_args(devices: Sequence[xb.xla_client.Device],
         buffers[r][a] = (buf if buf.device() == devices[r]
                          else buf.copy_to_device(devices[r]))
     else:
+      arg = xla.canonicalize_dtype(arg)
```
Comment (skye) [inline jax/interpreters/pxla.py:199]: Why did you need this? Maybe we should just overload the necessary handler more? (There's a TODO to get rid of canonicalize_dtype but I don't know anything about it)

File: tests/pmap_test.py [hunk 2]
```diff
@@ -208,6 +208,34 @@ def testNestedShardingAndStacking(self, device_mesh_shape):
...
+    f = pmap(lambda x, y: x - lax.psum(y, 'i'), axis_name='i', in_axes=(None, 0))
+    f = pmap(f, axis_name='j', in_axes=(None, 0))
+
+    x = 3.
+    y = onp.arange(prod(mesh_shape), dtype=onp.float32).reshape(mesh_shape)
+    expected = onp.broadcast_to(x - onp.sum(y, 1, keepdims=True), mesh_shape)
+
+    ans = f(x, y)
+    self.assertAllClose(ans, expected, check_dtypes=False)
+
```
Comment (skye) [inline tests/pmap_test.py:238]: Consider adding a test that returns a replicated parameter (and test that the sharding spec is not replicated), just because it's kind of an interesting edge case. I'm pretty sure it works, it would more be to demonstrate the behavior, and maybe we'd want to change it someday.

**Review by skye — APPROVED** [review]
Love it! It's pretty simple.

File: jax/interpreters/pxla.py (hunk 1 above)
Comment (jekbradbury) [reply jax/interpreters/pxla.py:199]: With this change, users can now end up passing scalars to pmap; that wasn't handled before because it couldn't happen at top level. You're right that we could add a scalar handler instead; I'm following xla.py here but I didn't realize we wanted to kill `canonicalize_dtype`. Think I should switch to a handler?

(Similarly, the change to `spec_to_indices` is because we are now calling that on specs that have scalar avals).

File: tests/pmap_test.py (hunk 2 above)
Comment (jekbradbury) [reply tests/pmap_test.py:238]: Good point, will add that! And I don't think we want to change the behavior, unless the user passes `out_axes=None`.

File: jax/interpreters/pxla.py (hunk 1 above)
Comment (skye) [reply jax/interpreters/pxla.py:199]: Ah makes sense. I would say keep it as-is, I'm not totally sure what the TODO is proposing anyway so may as well be consistent for now.

File: tests/pmap_test.py (hunk 2 above)
Comment (skye) [reply tests/pmap_test.py:238]: Good point that we probably shouldn't change the default behavior, woo.
